### PR TITLE
Overhaul literal infrastructure

### LINKED ISSUE #598: Leftover of #589
+ [x] In case `cons` constructor has implicit arguments, we should generate the correct implicitness accordingly. I think @HoshinoTented can handle this part.
+ [ ] I think the infrastructure of `Shaped` can be improved. Currently it's a bit wacky as we use `state` from time to time and it is nullable. I think we can do this: there is a thing "unknown type literal" that bans `constructorForm` and the type is a hole, and "known type literal" which does not need a `state` in `constructorForm` and the type is `DataCall`. When we need to use `constructorForm` of "unknown type literal" in a type-checking function, we try to convert it into known type literal by using the `state` and in this case, it's always NotNull. If it fails we can report error. This makes more operations type-safe and less error-prone. I think this should go to @imkiva 

See #589

### PR DESCRIPTION
**Pull request by imkiva**
Fix #598.

tl;dr: The unification now starts to play a role in solving literal types and disambiguation. RIP the old hacky and tricky code!

Details: When a literal may be represented by more than one datatype, we construct a hole for it and store it as the type in a `MetaLit`. Every time a _type-solved literal_ is required during unification, we eagerly check for the solution to the type and attempt to inline the `MetaLit` (like, when unfolding a global definition, normalizing a substituted term, etc.).

What's left: Because we need to know which universe list elements live in, which is only known when the underlying `DataCall` is provided, ambiguous list literals cannot yet be handled in this manner.

### CODE REVIEW

**Comment by ice1000** [comment]
Conflicts

**Comment by ice1000** [comment]
@imkiva conflict again

**Comment by imkiva** [comment]
little task: cover the code below

image

**Comment by ice1000** [comment]
Rest lgtm

File: base/src/main/java/org/aya/core/repr/AyaShape.java [hunk 1]
```diff
@@ -36,15 +40,17 @@ record AyaIntLitShape() implements AyaShape {
   }
 
   record AyaListShape() implements AyaShape {
+    public static final @NotNull CodeShape.MomentId NIL = new CodeShape.MomentId();
+    public static final @NotNull CodeShape.MomentId CONS = new CodeShape.MomentId();
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/repr/AyaShape.java:44]: Why not just use the all-in-one enum directly?

File: base/src/main/java/org/aya/core/repr/CodeShape.java [hunk 2]
```diff
@@ -29,8 +38,9 @@ record StructShape(
   ) implements CodeShape {}
 
   record CtorShape(
+    @NotNull MomentId name,
     @NotNull ImmutableSeq<ParamShape> tele
-  ) implements CodeShape {}
+  ) implements CodeShape, Moment {}
```
[the remark this answers is not in the document]
Comment (ice1000) [reply base/src/main/java/org/aya/core/repr/CodeShape.java:43]: Ok I see I was stupid

File: base/src/main/java/org/aya/core/term/MetaLitTerm.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import org.aya.core.def.GenericDef;
+import org.aya.core.repr.AyaShape;
+import org.aya.core.repr.ShapeRecognition;
+import org.aya.util.error.SourcePos;
+import org.jetbrains.annotations.NotNull;
+
+public record MetaLitTerm(
+  @NotNull SourcePos sourcePos,
+  @NotNull Object repr,
+  @NotNull ImmutableSeq<Tuple2<GenericDef, ShapeRecognition>> candidates,
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/term/MetaLitTerm.java:17]: ImmutableMap?

File: base/src/main/java/org/aya/core/term/MetaLitTerm.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import kala.collection.immutable.ImmutableSeq;
+import kala.control.Option;
+import kala.tuple.Tuple2;
+import org.aya.core.def.GenericDef;
+import org.aya.core.repr.AyaShape;
+import org.aya.core.repr.ShapeRecognition;
+import org.aya.util.error.SourcePos;
+import org.jetbrains.annotations.NotNull;
+
+public record MetaLitTerm(
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/term/MetaLitTerm.java:14]: Is there a "metavar" for metalits?

File: base/src/main/java/org/aya/core/term/MetaLitTerm.java (hunk 3 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/core/term/MetaLitTerm.java:17]: I think it is possible that the same data matches different shapes.
Maybe `ImmutableMap<GenericDef, ImmutableSeq<ShapeRecognition>>` ?

Comment (ice1000) [reply base/src/main/java/org/aya/core/term/MetaLitTerm.java:17]: I doubt. This represents the "candidates" which I do not expect to have so many.

But also since it's just a list of candidates, I tend to agree with the current code

File: base/src/main/java/org/aya/core/term/MetaLitTerm.java (hunk 4 above)
Comment (imkiva) [reply base/src/main/java/org/aya/core/term/MetaLitTerm.java:14]: `MetaLit` reuses the `MetaTerm` as its type, so the metavar is just ordinary meta (unlike `MetaPat`)

File: base/src/main/java/org/aya/core/repr/AyaShape.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/core/repr/AyaShape.java:44]: Good point

**Review by HoshinoTented — APPROVED** [review]

File: base/src/main/java/org/aya/core/term/MetaLitTerm.java (hunk 4 above)
Comment (ice1000) [reply base/src/main/java/org/aya/core/term/MetaLitTerm.java:14]: Cooooooooooooooool
